lightning-block-sync/src/poll.rs

```diff
@@ -0,0 +1,356 @@
+	Worse(ValidatedBlockHeader),
+}
+
+/// The `Validate` trait defines behavior for validating chain data.
+pub(crate) trait Validate {
+	/// The validated data wrapper which can be dereferenced to obtain the validated data.
+	type T: std::ops::Deref<Target = Self>;
+
+	/// Validates the chain data against the given block hash and any criteria needed to ensure that
+	/// it is internally consistent.
+	fn validate(self, block_hash: BlockHash) -> BlockSourceResult<Self::T>;
+}
+
+impl Validate for BlockHeaderData {
+	type T = ValidatedBlockHeader;
+
+	fn validate(self, block_hash: BlockHash) -> BlockSourceResult<Self::T> {
+		self.header
+			.validate_pow(&self.header.target())
+			.or_else(|e| Err(BlockSourceError::persistent(e)))?;
+
+		// TODO: Use the result of validate_pow instead of recomputing the block hash once upstream.
+		if self.header.block_hash() != block_hash {
+			return Err(BlockSourceError::persistent("invalid block hash"));
+		}
+
+		Ok(ValidatedBlockHeader { block_hash, inner: self })
+	}
+}
+
+impl Validate for Block {
+	type T = ValidatedBlock;
+
+	fn validate(self, block_hash: BlockHash) -> BlockSourceResult<Self::T> {
+		self.header
+			.validate_pow(&self.header.target())
+			.or_else(|e| Err(BlockSourceError::persistent(e)))?;
+
+		// TODO: Use the result of validate_pow instead of recomputing the block hash once upstream.
+		if self.block_hash() != block_hash {
+			return Err(BlockSourceError::persistent("invalid block hash"));
+		}
+
+		if !self.check_merkle_root() {
+			return Err(BlockSourceError::persistent("invalid merkle root"));
+		}
+
+		if !self.check_witness_commitment() {
+			return Err(BlockSourceError::persistent("invalid witness commitment"));
+		}
+
+		Ok(ValidatedBlock { block_hash, inner: self })
+	}
+}
+
+/// A block header with validated proof of work and corresponding block hash.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct ValidatedBlockHeader {
+	pub(crate) block_hash: BlockHash,
+	inner: BlockHeaderData,
+}
+
+impl std::ops::Deref for ValidatedBlockHeader {
+	type Target = BlockHeaderData;
+
+	fn deref(&self) -> &Self::Target {
+		&self.inner
+	}
+}
+
+impl ValidatedBlockHeader {
+	/// Checks that the header correctly builds on previous_header: the claimed work differential
+	/// matches the actual PoW and the difficulty transition is possible, i.e., within 4x.
+	fn check_builds_on(&self, previous_header: &ValidatedBlockHeader, network: Network) -> BlockSourceResult<()> {
+		if self.header.prev_blockhash != previous_header.block_hash {
+			return Err(BlockSourceError::persistent("invalid previous block hash"));
+		}
+
+		if self.height != previous_header.height + 1 {
+			return Err(BlockSourceError::persistent("invalid block height"));
+		}
+
+		let work = self.header.work();
+		if self.chainwork != previous_header.chainwork + work {
+			return Err(BlockSourceError::persistent("invalid chainwork"));
+		}
+
+		if let Network::Bitcoin = network {
+			if self.height % 2016 == 0 {
```

Comment:
I'm kinda surprised difficulty adjustment calculations are part of the sample module. This is certainly something I wouldn't expect third parties to replicate or copy and paste. Ought we create a utility for this? Also, third parties might be much more likely to use external dependencies, and those might have this covered.

Reply:
`ChainPoller` is such a utility. In fact, the only we to get validated data is to use it because the `Validate` trait is `pub(crate)`. As such, anyone implementing `Poll` must do so in terms of `ChainPoller`.